### ISSUE TITLE
Struct ssl_st cannot reference to s3 variables. Show an error  invalid use of incomplete type ‘const SSL {aka const struct ssl_st}’   ssl->s3;.

### ISSUE THREAD

**Issue by rchatsiri** [ripple/rippled#1914]
Hello contributor,
I cannot build rippled core engine with Make step cause build tools show an error as below
```
workspacecpp/rippled/src/ripple/basics/impl/make_SSLContext.cpp:348:5: error: invalid use of incomplete type ‘const SSL {aka const struct ssl_st}’
  ssl->s3;
     ^~
In file included from /usr/local/include/openssl/crypto.h:31:0,
                 from /usr/local/include/openssl/bio.h:20,
                 from /usr/local/include/openssl/conf.h:13,
```
Pointer to constant SSL  at member funciton name ```info_handler(SSL const * ssl, int event, int)```  should reference to [s3](http://docs.huihoo.com/doxygen/openssl/1.0.1c/include_2openssl_2ssl3_8h_source.html#l00405), but  represent an error  ``` invalid use of incomplete type ‘const SSL {aka const struct ssl_st}` ```. How to work around  is issue?  I build source code completed by comments declare variable at ```info_handler()``` member function.

**Comment by nbougalis** [comment]
Thanks for the report. What platform are you building on, and with what compiler? If on Linux, which distribution, and what is the version of OpenSSL you are using?

**Comment by rchatsiri** [comment]
Hello @nbougalis , I use  Linux debian-dev-00 3.16.0-4-amd64 #1 SMP Debian 3.16.36-1+deb8u2 (2016-10-19) x86_64 GNU/Linux. OpenSSL branch is  OpenSSL_1_0_2-stable cause  Rippled Cmake requires OpenSSL version 1.0.1

**Comment by vinniefalco** [comment]
Maybe we need to add `#include <ssl/ssl3.h>` or perhaps we have to set some macro to get that declaration.

**Comment by rchatsiri** [comment]
I think in the master branch of Ripple supports BoringSSL instead of OpenSSL library. After built the source code in the current master branch with OpenSSL show error at variable ```SSL_R_SHORT_READ```.  SSL_R_SHORT_READ replaced by  SSL_R_UNEXPECTED_RECORD variable in  ```is_short_read()``` of ```ssl_error.h``` file.  

Ref : https://github.com/chriskohlhoff/asio/issues/52

**Comment by nbougalis** [comment]
@rchatsiri: interesting, thanks for the link. Ripple doesn't use BoringSSL. I will sort this issue and get a patch. I'll keep you up to date.

**Comment by vinniefalco** [comment]
Is this still an issue?

**Comment by MarkusTeufelberger** [comment]
Both errors occur to me with OpenSSL 1.1 (seee #2047). Are you sure you're using 1.0.2?

@nbougalis - please check out if you can reproduce this with OpenSSL 1.1, it is getting into more and more distributions these days and rippled currently fails building with these 2 errors (short read and s3).

**Comment by MarkusTeufelberger** [comment]
Anything else you might need to reproduce this or that could help moving this along? This bug is blocking rippled from being compiled on at least every distribution that has made the switch to OpenSSL 1.1.

**Comment by MarkusTeufelberger** [comment]
@rchatsiri https://github.com/ripple/rippled/pull/2151 should have fixed this, does this still happen with 0.70.1?

**Comment by rchatsiri** [comment]
I change OpenSSL version to 1.0.2g 1 Mar 2016. it was bulit completed.
